### PR TITLE
Taking screenshot without colour filter

### LINKED ISSUE #1628: Taking a screenshot should bypass the colour filter
If a colour filter is active it will show up in screenshots and also when saving the game.
Would be nice for the screenshot to have normal colour.

### PR DESCRIPTION
**Pull request by takko**
**Brief Description of What This PR Does**

I switch off the colour filter, skip 2 frames, take the screenshot and switch the colour filter on.
A short screen blink is visible. So not sure if this is accepted.

This PR does some stuff...

**Related Issues**

closes #1628

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author.

### CODE REVIEW

**Review by Kasterisk — CHANGES_REQUESTED** [review]

**Review by FuJa0815 — APPROVED** [review]

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [inline src/engine/ScreenShotTaker.cs:57]:
```suggestion
    private void SaveScreenshot(Image image)
```

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [inline src/engine/ScreenShotTaker.cs:86]: Shouldn't this be in the SaveScreenshot method?

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [inline src/engine/ScreenShotTaker.cs:104]: Is it actually needed to wait this frame?

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [inline src/engine/ScreenShotTaker.cs:126]: Why is the image not written to disk when it was actually taken? Instead an extra frame is waited?

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [inline src/engine/ScreenShotTaker.cs:112]: These unnecessary prints should be removed.

File: src/saving/InProgressSave.cs
Comment (hhyyrylainen) [inline src/saving/InProgressSave.cs:182]: Please add a blank line before this to separate out that this is not related to the `save` variable initialization.

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [inline src/engine/ScreenShotTaker.cs:20]:
```suggestion
    private enum Step
```
I'd name this like that so that `Step.Start` reads more naturally to me.

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [inline src/engine/ScreenShotTaker.cs:47]:
```suggestion
    public Image GetViewportTextureAsImage()
```
Or if you prefer shorter names I think `GetViewportAsImage` would be perfectly descriptive as well.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/engine/ScreenShotTaker.cs
Comment (takko) [reply src/engine/ScreenShotTaker.cs:104]: Yes, it's needed to wait 2 frames. Tested it, and Godot doc says the same.

File: src/engine/ScreenShotTaker.cs
Comment (takko) [reply src/engine/ScreenShotTaker.cs:126]: if I save it directly the blink is much longer, as the save blocks the thread until save finished.
Saving in the next frame has the ScreenFilter on again.

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [reply src/engine/ScreenShotTaker.cs:105]: But I'm talking about waiting the one extra frame afterwards? I guess it doesn't matter if the extra frame *only* prevents another screenshot from happening. I was thinking that it would be bad if it prevented the game from working for one extra frame, but I guess that isn't the case after all.

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [reply src/engine/ScreenShotTaker.cs]: See my previous comment about the extra frame, so I guess it would be fine. What about writing the file to disk from a background task in TaskExecutor? I would expect it to be a thread safe operation of writing the file to disk.

Comment (takko) [reply src/engine/ScreenShotTaker.cs]: great idea, i try that.

Comment (takko) [reply src/engine/ScreenShotTaker.cs]: changed the last step to a background task.

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [inline src/engine/ScreenShotTaker.cs:95]: I'd refactor this bit so that you pass in the image to this lambda, and the lambda body would call an instance method on this object, that way there isn't a need to have separate variables to hold this task or the screenshot image.

**Review by hhyyrylainen — APPROVED** [review]
Confirmed functionality and I hope you don't mind me doing a bit of refactoring on this.

**Comment by hhyyrylainen** [comment]
I would usually wait before merging if I make changes, but this time I just want to get some PRs off my review list.

**Pull request merged**
